python/hackathon_app/sheets.py

```diff
@@ -21,121 +25,176 @@ def __init__(self, spreadsheet_id: str, cred_file: str):
+            serialized = self._convert_to_list(serialized_)
+            body = {"values": [serialized]}
+            self.client.append(
+                spreadsheetId=self.spreadsheet_id,
+                range=self.range,
+                insertDataOption="INSERT_ROWS",
+                valueInputOption="RAW",
+                body=body,
+            ).execute()
+        except:
+            print("Oops. No go.")
+
+    def rows(self) -> TStructure:
+        """Retrieve rows from sheet"""
+        try:
+            response = self.client.get(
+                spreadsheetId=self.spreadsheet_id, range=self.range
+            ).execute()
+            rows = response["values"]
+            data = self._convert_to_dict(rows)
+            response = cattr.structure(data, self.structure)
+        except (TypeError, AttributeError):
+            raise DeserializeError("Bad Data")
+        return response
+
+    def _convert_to_dict(self, data) -> Sequence[Mapping[str, str]]:
+        """Given a list of lists where the first list contains key names, convert it to
+        a list of dictionaries.
+        """
+        result: Sequence[Dict[str, str]] = [dict(zip(data[0], r)) for r in data[1:]]
+        return result
 
-    def create_user(
+    def _convert_to_list(
         self,
-        first_name: str,
-        last_name: str,
-        email: str,
-        company: str,
-        country: str,
-        tshirt_size: str,
-    ):
-        """Append user details to users sheet."""
-        data = {
-            "values": [
-                [
-                    first_name,
-                    last_name,
-                    email,
-                    company,
-                    country,
-                    datetime.now().strftime("%m/%d/%Y"),
-                    tshirt_size,
-                ]
-            ]
-        }
-        request = self.spreadsheet.append(
-            spreadsheetId=self.id,
-            range="users!A1:END",
-            insertDataOption="INSERT_ROWS",
-            valueInputOption="RAW",
-            body=data,
-        )
-        request.execute()
+        data: Mapping[str, Union[str, int, Sequence[str], datetime.datetime, None]],
+    ) -> Sequence:
+        """Given a dictionary, return a list containing its values"""
+        return list(data.values())
+
+
+@attr.s(auto_attribs=True, kw_only=True)
+class User:
+    first_name: str
+    last_name: str
+    email: str
+    date_created: Optional[datetime.datetime]
+    organization: str
+    tshirt_size: str
 
-    def _is_registered(self, email: str, hackathon_name: str):
-        """Check if user is already registed for a given hackathon."""
-        registrants = self.spreadsheet.get(
-            spreadsheetId=self.id, range="hackathons_users!A1:end"
-        ).execute()
+
+class Users(WhollySheet):
+    def __init__(self, client, spreadsheet_id):
+        super().__init__(client, spreadsheet_id, "users", Sequence[User])
```

Comment:
I think it would be better to pass `structure=User` and have `rows()` just wrap it in a `Sequence` (if that's possible, I'll find out)